<a href="https://colab.research.google.com/github/Dharma-Ranganathan/AllAboutPython/blob/main/DA_Projects_Codes/Grid_Search_CV.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Grid search cross validation is used to find hyperparameters of a model by tuning it.

Dependencies

In [135]:
import pandas as pd
import numpy as np
from sklearn.model_selection import GridSearchCV

Data collection

In [136]:
breast_cancer = pd.read_csv('/content/drive/MyDrive/Colab_python/breast_cancer_data.csv')

# checking
breast_cancer.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [137]:
print(breast_cancer)

           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compa

In [138]:
breast_cancer.tail()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
564  926424         M        21.56         22.39          142.00     1479.0   
565  926682         M        20.13         28.25          131.20     1261.0   
566  926954         M        16.60         28.08          108.30      858.1   
567  927241         M        20.60         29.33          140.10     1265.0   
568   92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
564                  0.07115          NaN  
565                  0.06637          NaN  
566                  0.07820          NaN  
567                  0.12400          NaN  
568                  0.07039          NaN  

[5 rows x 33 columns]

Something at last column is NaN , we need to drop that.

In [139]:
# dropping last column and id column
breast_cancer = breast_cancer.dropna(axis=1)
breast_cancer = breast_cancer.drop('id',axis=1)

# checking
breast_cancer.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

Data Pre-processing

In [140]:
# checking null values
breast_cancer.isnull().sum()

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

* no null values apart from last column we dropped

In [141]:
# checking shape
breast_cancer.shape

(569, 31)

In [142]:
# checking balanced state
breast_cancer.value_counts('diagnosis')

diagnosis
B    357
M    212
Name: count, dtype: int64

In [143]:

# benign

benign = breast_cancer['diagnosis'] == 'B'
benign = benign.sum()
benign

np.int64(357)

In [144]:
# malignant

malignant = breast_cancer['diagnosis'] == 'M'
malignant = malignant.sum()
malignant

np.int64(212)

* Total data points = 569
* Benign data points = 357
* Malignant data points = 212

In [145]:
# checking proportionality of diagnosis columns

benign_prop = (benign / breast_cancer.shape[0]) * 100
malignant_prop = (malignant/breast_cancer.shape[0]) * 100

print(round(benign_prop,2),'%', round(malignant_prop,2),'%')

62.74 % 37.26 %


* our dataset is not in that much Imbalanced, more or less is near to Imbalanced but no problem. We can proceed

Splitting features and labels(diagnosis)

In [146]:
features = breast_cancer.drop('diagnosis',axis=1)
label = breast_cancer.diagnosis

print('features')
print(features)
print()

print('label')
print(label)
print()

features
     radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     compactness_mean  concavity_mean  concave points_

Label Encoding using LabelEncoder class

In [147]:
# label encoding
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
label = encoder.fit_transform(label)
print(label)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 0 1 1 1 1 1 1 1 1 0 1 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 1
 0 1 0 1 1 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0
 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1 1 0 1 1 0 0 1 0 0 1 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 1 1 1 0 1
 0 1 0 0 0 1 0 0 1 1 0 1 1 1 1 0 1 1 1 0 1 0 1 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0
 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 1 0 0
 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1
 1 0 1 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 1 0 0 0 0 0 1 0 0
 1 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 1 0 1 0 0 1 

Converting features and label into numpy arrays for grid search cross validation

In [148]:
# as numpy arrays
x = np.asarray(features.values)
y = np.asarray(label)

print(x)
print()
print(y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 0 1 1 1 1 1 1 1 1 0 1 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 1
 0 1 0 1 1 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0
 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1 1 0 1 1 0 0 1 0 0 1 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 1 1 1 0 1
 0 1 0 0 0 1 0 0 1 1 0 1 1 1 1 0 1 1 1 0 1 0 1 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0

Implementation of grid search cv

In [149]:
# creating param_grid for grid search cv
params = { 'kernel' : ['kernel','poly','rbf','sigmoid'], 'C' : [1,5,10,20] }

In [150]:
# model svm
from sklearn.svm import SVC
model = SVC()

In [151]:
classifier = GridSearchCV(model,params,cv=5)

Providing x and y in classifier

In [ ]:
classifier.fit(x,y)

In [153]:
# results of classifier
classifier.cv_results_

{'mean_fit_time': array([0.000631  , 0.00379825, 0.00490851, 0.01512775, 0.00027981,
        0.00363207, 0.00454187, 0.01585808, 0.00032353, 0.0044426 ,
        0.00402017, 0.01301737, 0.00031986, 0.004252  , 0.00414891,
        0.01318774]),
 'std_fit_time': array([3.18475486e-04, 1.55511723e-04, 4.00978102e-04, 8.77023378e-04,
        2.30833514e-05, 1.77375668e-04, 6.35157960e-04, 3.23022319e-03,
        2.25796367e-05, 6.79191399e-04, 1.21308666e-04, 2.69267093e-04,
        6.93306470e-05, 1.99961047e-04, 3.59749901e-04, 4.57682766e-04]),
 'mean_score_time': array([0.        , 0.00145793, 0.00210729, 0.00398841, 0.        ,
        0.00135083, 0.00191684, 0.0045083 , 0.        , 0.0014585 ,
        0.0018033 , 0.00376825, 0.        , 0.00132074, 0.00167713,
        0.00374198]),
 'std_score_time': array([0.00000000e+00, 1.81968207e-05, 6.08927959e-05, 8.52359725e-05,
        0.00000000e+00, 4.80322971e-05, 7.13330539e-05, 6.47641317e-04,
        0.00000000e+00, 1.09356196e-04, 6.73

In [154]:
# best params we got

best_params = classifier.best_params_
print(best_params)

{'C': 5, 'kernel': 'rbf'}


In [155]:
# cv results to dataframe

results = pd.DataFrame(classifier.cv_results_)
results.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_C  \
0       0.000631      0.000318         0.000000        0.000000        1   
1       0.003798      0.000156         0.001458        0.000018        1   
2       0.004909      0.000401         0.002107        0.000061        1   
3       0.015128      0.000877         0.003988        0.000085        1   
4       0.000280      0.000023         0.000000        0.000000        5   

  param_kernel                         params  split0_test_score  \
0       kernel   {'C': 1, 'kernel': 'kernel'}                NaN   
1         poly     {'C': 1, 'kernel': 'poly'}           0.842105   
2          rbf      {'C': 1, 'kernel': 'rbf'}           0.850877   
3      sigmoid  {'C': 1, 'kernel': 'sigmoid'}           0.543860   
4       kernel   {'C': 5, 'kernel': 'kernel'}                NaN   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0                NaN                NaN                NaN                NaN   
1           0.885965           0.929825           0.947368           0.938053   
2           0.894737           0.929825           0.947368           0.938053   
3           0.456140           0.464912           0.385965           0.451327   
4                NaN                NaN                NaN                NaN   

   mean_test_score  std_test_score  rank_test_score  
0              NaN             NaN               13  
1         0.908663        0.039382                8  
2         0.912172        0.035444                7  
3         0.460441        0.050253                9  
4              NaN             NaN               13

Creating a new data frame out of this results data frame

In [162]:
insight = results[['param_C','param_kernel','mean_test_score','rank_test_score']]

print(insight)
print()

    param_C param_kernel  mean_test_score  rank_test_score
0         1       kernel              NaN               13
1         1         poly         0.908663                8
2         1          rbf         0.912172                7
3         1      sigmoid         0.460441                9
4         5       kernel              NaN               13
5         5         poly         0.922729                2
6         5          rbf         0.931501                1
7         5      sigmoid         0.411178               10
8        10       kernel              NaN               13
9        10         poly         0.920975                4
10       10          rbf         0.922714                3
11       10      sigmoid         0.402391               11
12       20       kernel              NaN               13
13       20         poly         0.919221                6
14       20          rbf         0.920944                5
15       20      sigmoid         0.398867               

We got to know , rank 1 has kernel 'rbf' and C=5

Using this hyperparameters, we can develop a new model